File: apps/desktop/src/about/githubRelease.ts
import { GITHUB_RELEASES_API } from './donate.ts'
import type { NamedAsset } from './productUpdate.ts'

export interface GithubRelease {
  htmlUrl: string
  assets: NamedAsset[]
}

export function parseGithubRelease(raw: unknown): GithubRelease | null {
  if (!raw || typeof raw !== 'object') return null
  const obj = raw as { html_url?: unknown; assets?: unknown }
  const htmlUrl = typeof obj.html_url === 'string' ? obj.html_url : ''
  if (!Array.isArray(obj.assets)) return { htmlUrl, assets: [] }
  const assets: NamedAsset[] = []
  for (const item of obj.assets) {
    if (!item || typeof item !== 'object') continue
    const a = item as { name?: unknown; browser_download_url?: unknown }
    if (typeof a.name === 'string' && typeof a.browser_download_url === 'string') {
      assets.push({ name: a.name, url: a.browser_download_url })
    }
  }
  return { htmlUrl, assets }
}

export async function fetchLatestGithubRelease(
  currentVersion: string,
  fetchImpl: typeof fetch = fetch,
): Promise<GithubRelease | null> {
  const ctrl = new AbortController()
  const timer = setTimeout(() => ctrl.abort(), 10_000)
  try {
    const res = await fetchImpl(GITHUB_RELEASES_API, {
      signal: ctrl.signal,
      headers: {
        Accept: 'application/vnd.github+json',
        'User-Agent': `Continuum-Calendar/${currentVersion}`,
      },
    })
    if (!res.ok) return null
    return parseGithubRelease(await res.json())
  } catch {
    return null
  } finally {
    clearTimeout(timer)
  }
}

File: examples/android/app/src/main/java/dev/foss/goldenpath/about/GithubRelease.kt
package dev.foss.goldenpath.about

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

object GithubRelease {
    data class Parsed(val htmlUrl: String, val assets: List<ProductUpdate.NamedAsset>)

    fun parse(json: String, fallbackHtml: String = ""): Parsed? {
        return try {
            val root = JSONObject(json)
            val htmlUrl = root.optString("html_url", fallbackHtml)
            val assets = mutableListOf<ProductUpdate.NamedAsset>()
            val arr = root.optJSONArray("assets") ?: return Parsed(htmlUrl, assets)
            for (i in 0 until arr.length()) {
                val item = arr.optJSONObject(i) ?: continue
                val name = item.optString("name")
                val url = item.optString("browser_download_url")
                if (name.isNotBlank() && url.isNotBlank()) {
                    assets.add(ProductUpdate.NamedAsset(name, url))
                }
            }
            Parsed(htmlUrl, assets)
        } catch (_: Exception) {
            null
        }
    }

    fun fetchLatest(apiUrl: String, userAgent: String): Parsed? {
        if (apiUrl.isBlank()) return null
        val conn = URL(apiUrl).openConnection() as HttpURLConnection
        return try {
            conn.requestMethod = "GET"
            conn.setRequestProperty("Accept", "application/vnd.github+json")
            conn.setRequestProperty("User-Agent", userAgent)
            conn.connectTimeout = 10_000
            conn.readTimeout = 10_000
            if (conn.responseCode != HttpURLConnection.HTTP_OK) return null
            parse(conn.inputStream.bufferedReader().use { it.readText() })
        } catch (_: Exception) {
            null
        } finally {
            conn.disconnect()
        }
    }
}
